### PR TITLE
use double hashing instead of multiple hashes

### PR DESCRIPTION
**Pull request by funny-falcon**
See https://www.eecs.harvard.edu/~michaelm/postscripts/rsa2008.pdf
Also use hash function which produce couple of hashsums for double hashing
at once.

### CODE REVIEW

File: src/bloom_filter/filter.cr [hunk 1]
```diff
@@ -10,8 +8,8 @@ module BloomFilter
       @hash_num = hash_num.to_u8
 
       @bitmap = Array(UInt8).new(bytesize, 0_u8)
-      @seeds = [] of UInt32
-      setup_seeds
+      @seeda = 0xdeadbeef_u32
+      @seedb = 0x71fefeed_u32
```
Comment (greyblake) [inline src/bloom_filter/filter.cr]: Those should be extracted as constants

File: src/bloom_filter/filter.cr [hunk 2]
```diff
@@ -29,22 +27,20 @@ module BloomFilter
       end
 
       @bitsize = bytesize * 8
-      @seeds = [] of UInt32
-      setup_seeds
+      @seeda = 0xdeadbeef_u32
+      @seedb = 0x71fefeed_u32
```
Comment (greyblake) [inline src/bloom_filter/filter.cr]: here as well

File: src/bloom_filter/filter.cr [hunk 3]
```diff
@@ -100,14 +90,36 @@ module BloomFilter
...
+      end
+    end
+
+    # Hash function.
+    private def two_hash(str : String) : Tuple(UInt32, UInt32)
+      ha = @seeda
+      hb = @seedb
+      str.each_byte do |b|
+        ha = (hswap(ha) ^ b) * 0xb8b34b2d_u32
+        hb = (hswap(hb) ^ b) * 0x52c6a2d9_u32
```
Comment (greyblake) [inline src/bloom_filter/filter.cr]: Wheret those magic numbers come from?

**Comment by greyblake** [comment]
@funny-falcon Thanks for the pull request! I took a brief look ,and left you comments. I'll take look at the paper tonight, to understand better the changes:)

File: src/bloom_filter/filter.cr (hunk 3 above)
Comment (funny-falcon) [inline src/bloom_filter/filter.cr]: From experiments with custom hash function https://github.com/funny-falcon/funny_hash/blob/master/funny_hash.h - they provides good avalanche when tested with SMHasher.

They are not "hardwired" constants: they should be odd and have some non regular bit pattern, but this pair is not "super magic".

File: src/bloom_filter/filter.cr (hunk 1 above)
Comment (funny-falcon) [inline src/bloom_filter/filter.cr]: I'll rather generate them and save/load to file, but it is your choice to do or not to do it.

**Comment by greyblake** [comment]
Benchmark (the current version):
```
  Size: 117005Kb
  Hash functions: 7

                     user     system      total        real
add              0.005630   0.000000   0.005630 (  3.687145)
has? (present)   0.016525   0.000000   0.016525 (  10.817915)
has? (missing)   0.001892   0.000000   0.001892 (  1.239114)
```
You version:
```
                     user     system      total        real
add              0.008652   0.000000   0.008652 (  5.668261)
has? (present)   0.008453   0.000000   0.008453 (  5.528132)
has? (missing)   0.004425   0.000015   0.004440 (  2.904493)
```
To be honest I expected to be it slowlier, but there is a nice bust in `has? (present)` ! =)
Thanks!

File: src/bloom_filter/filter.cr (hunk 1 above)
Comment (greyblake) [inline src/bloom_filter/filter.cr]: Is there any benefit from generating them dynamically?

**Comment by funny-falcon** [comment]
I didn't expect it is slower on `has?(missing)` and `add` :-( I'll play with.

**Pull request merged**